packages/host/app/services/render-service.ts

```diff
@@ -22,27 +48,121 @@ export default class RenderService extends Service {
+          isNotReadyError(e)
+        ) as NotReady | undefined;
+        notLoaded = err.additionalErrors?.find((e: any) =>
+          isNotLoadedError(e)
+        ) as NotLoaded | undefined;
+        if (isCardError(err) && (notReady || notLoaded)) {
+          let errorInstance = notReady?.instance ?? notLoaded?.instance;
+          if (!errorInstance) {
+            throw new Error(
+              `bug: could not determine NotLoaded/NotReady card instance`
+            );
+          }
+          let fieldName = notReady?.fieldName ?? notLoaded?.fieldName;
+          if (!fieldName) {
+            throw new Error(
+              `bug: could not determine NotLoaded/NotReady field`
+            );
+          }
+          await this.resolveField({
+            card: errorInstance,
+            fieldName,
+            identityContext,
+            visit,
+            realmPath,
+          });
+        } else {
+          throw err;
+        }
+      }
+    } while ((notLoaded || notReady) && tries++ <= maxRenderThreshold);
+
+    // This guards against bugs that may trigger render cycles
+    if (tries > maxRenderThreshold) {
+      let error = new CardError(
+        `detected a cycle trying to render card ${card.constructor.name} (id: ${card.id})`
+      );
+      error.additionalErrors = [(notLoaded ?? notReady)!];
+      throw error;
+    }
 
-    render(component, element, this.owner);
+    let serializer = new Serializer(voidMap);
     let html = serializer.serialize(element);
     return parseCardHtml(html);
   }
+
+  private async resolveField(
+    params: Omit<RenderCardParams, 'format'> & { fieldName: string }
+  ): Promise<void> {
+    let { card, visit, identityContext, realmPath, fieldName } = params;
+    let api = await this.loaderService.loader.import<typeof CardAPI>(
+      `${baseRealm.url}card-api`
+    );
+    let value: any;
+    try {
+      value = await api.getIfReady(card, fieldName as keyof Card, undefined, {
+        loadFields: true,
+      });
+    } catch (err: any) {
+      let notLoaded = err.additionalErrors?.find((e: any) =>
+        isNotLoadedError(e)
+      ) as NotLoaded | undefined;
+      if (isCardError(err) && notLoaded) {
+        let linkURL = new URL(`${notLoaded.reference}.json`);
+        if (realmPath.inRealm(linkURL)) {
+          await visit(linkURL, identityContext);
+        } else {
+          // in this case the instance we are linked to is a missing instance
+          // in an external realm.
+          throw err;
+        }
+      } else {
+        throw err;
+      }
+    }
+
+    if (Array.isArray(value)) {
+      for (let item of value) {
+        if (item && api.isCard(item)) {
+          await this.renderCard({
+            card: item,
+            visit,
+            realmPath,
+            identityContext,
+          });
+        }
+      }
+    }
+    if (api.isCard(value)) {
+      await this.renderCard({ card: value, visit, realmPath, identityContext });
+    }
```

Comment:
do we really need this? why are we having to recurse into the render card for each card here?